pkg/ingestor/ingestcli/client.go

```diff
@@ -0,0 +1,249 @@
+// Copyright 2025 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ingestcli
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/errorpb"
+	"github.com/pingcap/tidb/br/pkg/restore/split"
+	"github.com/pingcap/tidb/pkg/util"
+	"go.uber.org/atomic"
+)
+
+type nextGenResp struct {
+	SstMeta nextGenSSTMeta `json:"sst_meta"`
+}
+
+type nextGenSSTMeta struct {
+	Id         int64 `json:"id"`
+	Smallest   []int `json:"smallest"`
+	Biggest    []int `json:"biggest"`
+	MetaOffset int   `json:"meta-offset"`
+	CommitTs   int   `json:"commit-ts"`
+}
+
+var _ WriteClient = &writeClient{}
+
+type writeClient struct {
+	tikvWorkerURL string
+	clusterID     uint64
+	taskID        int64
+	httpClient    *http.Client
+	commitTS      uint64
+
+	eg         *util.ErrorGroupWithRecover
+	sendReqErr atomic.Error
+	writer     *io.PipeWriter
+	reader     *io.PipeReader
+	sstMeta    *nextGenSSTMeta
+}
+
+// newWriteClient creates a writeClient.
+func newWriteClient(
+	tikvWorkerURL string,
+	clusterID uint64,
+	httpClient *http.Client,
+	commitTS uint64,
+) *writeClient {
+	return &writeClient{
+		tikvWorkerURL: tikvWorkerURL,
+		clusterID:     clusterID,
+		commitTS:      commitTS,
+		httpClient:    httpClient,
+		eg:            util.NewErrorGroupWithRecover(),
+	}
+}
+
+func (w *writeClient) init(ctx context.Context) error {
+	pr, pw := io.Pipe()
+	url := fmt.Sprintf("%s/write_sst?cluster_id=%d&commit_ts=%d",
+		w.tikvWorkerURL, w.clusterID, w.commitTS)
+	req, err := http.NewRequestWithContext(ctx, "PUT", url, pr)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	req.Header.Set("Content-Type", "application/octet-stream")
+	w.startChunkedHTTPRequest(req)
+	w.reader = pr
```

Comment:
add some comment that `w.reader` will be closed by the `httpClient.Do` automatically, it's easy to forget the reason why we don't do it and think it's a bug